Initial checkin to support exporting Debian packages from Habitat.

yzl opened the pull request:
* Build packages within a studio started for the purpose.
* Install a minimum of packages to support building Debian packages.
* Support for dpkg-deb options is stolen from https://github.com/smith/hab-pkg-deb.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

cc @schisamo @smith

Commit:
Initial checkin to support exporting Debian packages from Habitat.

* Build packages within a studio started for the purpose.
* Install a minimum of packages to support building Debian packages.
* Support for dpkg-deb options is stolen from https://github.com/smith/hab-pkg-deb.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Look for preinst, postinst, prerm, and postrm scripts in the bin
directory of the package being exported.

Scripts specified via the CLI options --preinst, etc., will take
precedence.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+  core/sed
+  core/util-linux
+  core/dpkg
+    core/findutils
```
These don't need to be indented 4 spaces; only 2. The whole list should be alphabetized.

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+# This file is the heart of your application's habitat.
```
This comment can be removed.

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+# See full docs at https://www.habitat.sh/docs/reference/plan-syntax/
+
+pkg_name=hab-pkg-deb
+pkg_origin=yzl
```
Since this if chef/habitat-exporters, should this be `chef`? When building, your `$HAB_ORIGIN` will automatically override this if set.

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+pkg_upstream_url="https://github.com/chef/habitat-exporters"
+
+
+# Callback Functions
```
This comment can be removed

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+  return 0 
+}
+
+do_prepare() {
```
This does nothing by default and can be removed.

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,80 @@
+    -i $pkg_name
+}
+
+# The default implementation runs nothing during post-compile. An example of a
```
This comment and function can be removed.

smith reviewed (changes requested):
Looks great! There's a few things that don't need to be in the plan that can be removed, so let's get them out. Let me know if you have any questions.

schisamo commented on hab-pkg-deb/bin/hab-pkg-deb.sh:
```diff
@@ -0,0 +1,390 @@
+# The name converted to all lowercase to be compatible with Debian naming
+# conventions
+convert_name() {
+  safe_name="${pkg_origin,,}-${pkg_name,,}"
```
This lowercasing variable expansion is only available in Bash 4.0+. This shouldn't be an issue as we control the Bash version but we might want to leave a comment.

schisamo commented on hab-pkg-deb/bin/hab-pkg-deb.sh:
```diff
@@ -0,0 +1,390 @@
+Section: $(section)
+Priority: $(priority)
+EOF
+)
```
Can we render the control files off of a Handlebars template (same templating syntax the rest of Habitat uses). Since we control the deps for all of our exporters we could leverage a CLI (like [handlebars-cmd](https://github.com/spion/handlebars-cmd) or [hbs-cli](https://github.com/keithamus/hbs-cli) if we created a [core-plan](https://github.com/habitat-sh/core-plans) for it.

This approach would make all of the control file rendering code much more readable and maintainable. It also opens up the ability for an end-user to provide their own control file template.

schisamo reviewed (changes requested):
Let's go ahead and enable `shellcheck` linting on this repo and make sure it runs on all PRs via Travis (I [flipped the switch on Travis's side](https://travis-ci.org/chef/habitat-exporters)). Take a look at how we execute linting via pre-commit in the [habitat-sh/core-plans](https://github.com/habitat-sh/core-plans) repo:
https://github.com/habitat-sh/core-plans/blob/master/.travis.yml
https://github.com/habitat-sh/core-plans/blob/master/.pre-commit-config.yaml
https://github.com/habitat-sh/core-plans/blob/master/bin/travis.sh
https://github.com/habitat-sh/core-plans/blob/master/bin/check-default-variables.sh

Happy to walk you through all of this in person aslo @yzl.

Commit:
Clean up plan per @smith's review comments.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Set up shellcheck linting and fix errors.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Use handlebars template for control file.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Install dependencies to support use of handlebars templating.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Rework options handling to use handlebars templating. Also,

* If software being packaged supplies a conffiles list in export/deb/conffiles, it will be incorporated into the exported Debian package.
* If software being packaged supplies install scripts (preinst, etc.) in bin, it will be incorporated into the exported Debian package.
* The --depends, --section, and --priority options allow package dependencies, section, and priority to be supplied via the command-line.
* Under the hood, software being packaged is installed in a clean habitat studio and then appropriate pieces are staged in a separate directory structure.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

yzl commented:
@schisamo @smith I've made the following changes:

* Removed unnecessary code from my plan.sh.
* Used handlebars-cmd to render the control file.
* Enabled shellcheck linting.
* When install scripts are provided in the bin directory of the software being packaged, they will be incorporated into the .deb package. When a conffiles list is provided in the export/deb directory of the software being packaged, it will be incorporated into the .deb as well.

See individual commits for details.

schisamo commented on hab-pkg-deb/bin/hab-pkg-deb.sh:
```diff
@@ -0,0 +1,443 @@
+    --depends "$depends" \
+    --provides "$provides" \
+    --replaces "$replaces" \
+    < control/control.hbs \
```
Can an end user can provide there own `control/control.hbs`? Can this be provided at `export/deb/control.hbs`? We should also drop the `*.hbs` on this file. Idiomatically we don't use an extension like this on any other Handlebars templates in Habitat.

schisamo reviewed (changes requested):
Can we please get the README updated with some basic usage info? In particular we need to documen the various CLI options and what items you can provide in your package (stuff like `export/deb/conffiles`).

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,49 @@
+}
+
+do_install() {
+  install -v -D "$pkg_name" "$pkg_prefix/bin/$pkg_name"
```
We also need to copy in control/control.hbs otherwise I get:

```
/hab/pkgs/smith/hab-pkg-deb/0.1.0/20170411201942/bin/hab-pkg-deb: line 329: control/control.hbs: No such file or directory
```

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,49 @@
+  core/dpkg
+  core/findutils
+  core/hab
+  core/node
```
Do we depend directly on node or is this a transitive dependency of handlebars-cmd? If it's the latter we can remove it.

Commit:
A user can provide their own control file template as a file named `control`
in the `export/deb` directory of their package. If this file is not present,
hab-pkg-deb will use its own control file template.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
node is a transitive dependency and hence can be removed.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

smith commented on hab-pkg-deb/bin/hab-pkg-deb.sh:
```diff
@@ -0,0 +1,443 @@
+    --depends "$depends" \
+    --provides "$provides" \
+    --replaces "$replaces" \
+    < "$(hab pkg path yzl/hab-pkg-deb)/control/control" \
```
This does not work when I run this because that's not the origin under which I built this package.

You probably want to use a relative path here. Here's a (complex but reliable) way to get the directory of the script you're running (like `File.dirname(__FILE__)` in Ruby): https://gist.github.com/smith/3d32bb0b725d9cfb957cc23abde3eaae#get-the-directory-of-the-script-youre-running-in-bash

Commit:
Add inspec tests and supporting data.

* Refactor plan/install to enable testing via do_check().
* Add a testing mode that does not build packages and does keep staging directories around for further examination.
* Add tests.

Additionally
* Move control template to export/deb for consistency.
* Do not overload pkg_name in control file template.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Alphabetize options.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

schisamo commented:
@yzl Just want to make sure you saw: https://github.com/chef/habitat-exporters/pull/1#pullrequestreview-32121920

yzl commented:
@schisamo Yep -- I (finally) finished writing tests and am writing docs now.

Commit:
Add a proper readme.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

Commit:
Add --archive switch to allow full Debian archive name to be specified on the CLI. Also, populate Iteration field in the control file.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

yzl commented:
@schisamo I added documentation and tests, so this is ready for review again...

schisamo reviewed (approved):
This looks great @yzl. 

Let's move `hab-pkg-deb/tests/controls` into `hab-pkg-deb/tests/inspec/controls`. You'll also want to ensure you have an `inspec.yml` file at `hab-pkg-deb/tests/inspec/inspec.yml` (this should also move the generated `inspec.lock` to the same location). Take a look at https://github.com/chef/es-package-router/tree/master/inspec/package-router-smoke for an example.

yzl commented:
@schisamo I moved `tests/controls` to `tests/inspec/controls`, added `inspec.yml`, and checked in the `inspec.lock`.

schisamo commented:
@yzl The `inspec.lock` should actually live at `hab-pkg-deb/tests/inspec/inspec.lock`.

yzl commented:
@schisamo When I put the lock file in `hab-pkg-deb/tests/inspec/inspec.lock` and run the build/tests, a lock file gets regenerated in `hab-pkg-deb/tests/inspec/controls/inspec.lock`.

schisamo commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,60 @@
+
+do_check() {
+  "$PLAN_CONTEXT/tests/setup.sh" "$PWD" "$PLAN_CONTEXT"
+  hab pkg exec chef/inspec inspec exec "$PLAN_CONTEXT/tests/inspec/controls"
```
@yzl Your lock is being generated at `hab-pkg-deb/tests/inspec/controls/inspec.lock` because of the path on this `inspec exec` command. Try running:

```
hab pkg exec chef/inspec inspec exec "$PLAN_CONTEXT/tests/inspec"
```

This should properly create the `inspec.lock` at `hab-pkg-deb/tests/inspec/inspec.lock` alongside the `inspec.yml` file.

yzl commented:
@schisamo Fixed now.

Commit:
Move tests/controls to tests/inspec/controls for consistency with other projects.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

smith commented on hab-pkg-deb/plan.sh:
```diff
@@ -0,0 +1,60 @@
+
+do_check() {
+  "$PLAN_CONTEXT/tests/setup.sh" "$PWD" "$PLAN_CONTEXT"
+  hab pkg exec chef/inspec inspec exec "$PLAN_CONTEXT/tests/inspec"
```
If you have inspec as a build dep this can just be `inspec exec ...` since inspec will be on your PATH.

Commit:
Remove unneeded `hab pkg exec` per @smith.

Signed-off-by: Yvonne Lam <yvonne@chef.io>

yzl commented:
@schisamo @smith I think I've addressed your suggestions.  Are we good to merge?
